### DISCUSSION TITLE
Training on custom dataset

### DISCUSSION THREAD

**Discussion by anantgupta129** [deepdoctection/deepdoctection#59] [Q&A] [answered]
i want train the model for detection and extraction of tables. I am not sure on how the data should be labeled and the training procedure can you help me with that

**Comment by JaMe76** [comment, marked as answer]
Hi, thank you for your question.

Maybe you can have a look in the [discussion section](https://github.com/deepdoctection/deepdoctection/discussions/42): There is a chat about how to setup a custom dataset and how to transform data into the library framework. 

Feel free to ask any further questions that haven't been covered so far.

**Comment by anantgupta129** [comment]
> Hi, thank you for your question.
> 
> Maybe you can have a look in the [discussion section](https://github.com/deepdoctection/deepdoctection/discussions/42): There is a chat about how to setup a custom dataset and how to transform data into the library framework.
> 
> Feel free to ask any further questions that haven't been covered so far.

the table extractor is trained on pubtabnet, i downloaded the dataset. and checking the labels in jsonl file. Is that how i am supposed to label my dataset. If yes can you tell me which tool they are using for labelling

**Reply by JaMe76** [reply]
As far as I remember Pubtabnet has been annotated automatically. Check the official [Repo](https://github.com/ibm-aur-nlp/PubTabNet) for more information.

Generating labels and preparing labeled datasets for training models are different things. If you want to fine tune the object detectors to your custom table data you need to have labelled table data with bounding boxes for cells as well as rows and columns at hand. They can be in any raw format (COCO, Pascal-VOC, etc.) and you need to map them into the custom format. Check e.g. how Publaynet dataset has been implemented where ground truth data has been provided in COCO format which might be a good starting point.

**Reply by anantgupta129** [reply]
Thanks for the suggestion I will check how Publaynet dataset is labelled and label my data accordingly and use it for training deepdoctection for extraction of table. Might bother you again in future for help

Thanks

**Reply by anantgupta129** [reply]
Hi @JaMe76 
just need to clarify some points 
1. For training html data labeling is not required, it is generated internally 
2. Publaynet doesn't label table cell and rows level, i need to train model to identify table structure also then extraction, so i can label it like pubtabnet  
3. which dataset did you use to train table detection and extractor

**Reply by JaMe76** [reply]
I am not quite sure if I understand your concerns correctly. 

Publaynet is a dataset for object detection and is presented in COCO format. Annotations in COCO format can be used quite easily for feeding a model in train mode. 

Pubtabnet is a dataset for table recognition and is presented in a different and more elaborate format. It contains HTML as well as bounding boxes but can be used for training object detection models as well. It has bounding boxes for table cells and using the HTML representation and cell bounding boxes you can infer bounding boxes for rows and columns. This is what I do [here](https://github.com/deepdoctection/deepdoctection/blob/master/deepdoctection/mapper/pubstruct.py). The transformation, however is not trivial. 

I use Pubtabnet for training a cell detector and a second detector for rows and columns. Check [this page](https://deepdoctection.readthedocs.io/en/latest/tutorials/fine_tune_notebook.html) for an overview. After that I do a lot of post processing for assigning a row number/column number/row span/col span to each cell. Table recognition here does not work like that you have a table image, pass it to a model and then you get all results straight away. Check the [table recognition part of the analyzer](https://github.com/deepdoctection/deepdoctection/blob/master/deepdoctection/analyzer/dd.py) from line 158 up to 179: It involves 4 services (cell detection, row-col detection, table segmentation and table refinement).  

Saying that, it does not mean that you have to provide a dataset in Pubtabnet structure to fine tune models. You have to provide a dataset with table cell and row column bounding boxes to fine-tune the cell detector and the row/column detector. If you have to prepare a dataset from scratch anyway, you do not have to bother about creating ground truth in Pubtabnet format. Creating a dataset e.g. in COCO format will do the job and will certainly be easier. Having a dataset with annotations as given as in Pubtabnet is a sufficient condition to prepare training data for cell/row/column detection but it is not necessary. It contains much more information that you do not need, in order to fine models for solving a table recognition task following the approach here.
